### ISSUE TITLE
QUESTION: object_name_linter different depending on context?

### ISSUE THREAD

**Issue by minimenchmuncher** [jimhester/lintr#849]
I have a question about using the `object_name_linter` - is it possible for it to lint differently depending on the context - `snake_case` for example for variable names, `lowerCamelCase` for functions and `UpperCamelCase` for classes and class generators?

**Comment by AshesITR** [comment]
Currently this is not possible.
Note that distinguishing between the three can be hard without running the code, cf.

    library(R6)
    build_fun <- function() { function(...) 42 } # returns a function
    build_class <- function() { R6Class("test") }
    actually_a_function <- build_fun()
    actually_a_class <- build_class()
    R6Class <- build_fun()
    actually_a_variable <- build_class()

and so on...

**Comment by minimenchmuncher** [comment]
Point very well taken- this is absolutely a limitation in R being dynamically typed (generally); however, we should be able to get close by using regular expressions:

```r
library(R6)
library(methods)
## detect a function definition
grepl("(<-|=)[ ]+?function", code)

## detect a class definition
grepl("(<-|=)[` ]+?(R6Class|setClass)", code)

## detect a function which is a constructor
grepl("\\$?new\\(" code)
```

Admittedly, not too familiar with the `lintr` codebase and the `object_name_linter` - it looks like there are already far more sophisticated ways of doing checks off of the xml-parsed document than my (probably not-quite correct) regexes. The above are just sort of idea of a first cut of how it could be accomplished.

One thing I'll add, I noted the documentation says that most of the linters are based on [the tidyverse style guide, in this case here](https://style.tidyverse.org/syntax.html#object-names) however, [the documentation for R6 classes](https://r6.r-lib.org/articles/Introduction.html) and some of the [documentation for S4 classes](https://cran.r-project.org/doc/contrib/Genolini-S4tutorialV0-5en.pdf) (pg 15) prefer UpperCamelCase for class names and constructors, S4 seems to prefer lowerCamelCase for methods whereas R6 seems to prefer snake_case... but in either case, both paradigms would likely occur within the same package or file.

**Comment by AshesITR** [comment]
In my own packages, I use CamelCase for R6 classes as well, but they aren't so many, so using dev lintr's new feature to only exclude `object_name_linter` works for me.

e.g.

    MyClass <- R6::R6Class( # nolint: object_name_linter.
      "my_class",
      public = list(...)
    )
    
    my_class <- function(...) {
      MyClass$new(...)
    }

**Comment by AshesITR** [comment]
Crossref #1104
